pkg/kubelet/util/store/filestore.go

```diff
@@ -0,0 +1,157 @@
+	tmpPrefix = "."
+)
+
+// FileStore is an implementation of the Store interface which stores data in files.
+type FileStore struct {
+	// Absolute path to the base directory for storing data files.
+	directoryPath string
+
+	// filesystem to use.
+	filesystem utilfs.Filesystem
+}
+
+// NewFileStore returns an instance of FileStore.
+func NewFileStore(path string, fs utilfs.Filesystem) (Store, error) {
+	if err := ensureDirectory(fs, path); err != nil {
+		return nil, err
+	}
+	return &FileStore{directoryPath: path, filesystem: fs}, nil
+}
+
+// Write writes the given data to a file named key.
+func (f *FileStore) Write(key string, data []byte) error {
+	if err := ValidateKey(key); err != nil {
+		return err
+	}
+	if err := ensureDirectory(f.filesystem, f.directoryPath); err != nil {
+		return err
+	}
+
+	return writeFile(f.filesystem, f.getPathByKey(key), data)
+}
+
+// Read reads the data from the file named key.
+func (f *FileStore) Read(key string) ([]byte, error) {
+	if err := ValidateKey(key); err != nil {
+		return nil, err
+	}
+	bytes, err := f.filesystem.ReadFile(f.getPathByKey(key))
+	if os.IsNotExist(err) {
+		return bytes, ErrKeyNotFound
+	}
+	return bytes, err
+}
+
+// Delete deletes the key file.
+func (f *FileStore) Delete(key string) error {
+	if err := ValidateKey(key); err != nil {
+		return err
+	}
+	return removePath(f.filesystem, f.getPathByKey(key))
+}
+
+// List returns all keys in the store.
+func (f *FileStore) List() ([]string, error) {
+	keys := make([]string, 0)
+	files, err := f.filesystem.ReadDir(f.directoryPath)
+	if err != nil {
+		return keys, err
+	}
+	for _, f := range files {
+		if !strings.HasPrefix(f.Name(), tmpPrefix) {
+			keys = append(keys, f.Name())
+		}
+	}
+	return keys, nil
+}
+
+// getPathByKey returns the full path of the file for the key.
+func (f *FileStore) getPathByKey(key string) string {
+	return filepath.Join(f.directoryPath, key)
+}
+
+// ensureDirectory creates the directory if it does not exist.
+func ensureDirectory(fs utilfs.Filesystem, path string) error {
+	if _, err := fs.Stat(path); err != nil {
+		// MkdirAll returns nil if directory already exists.
+		return fs.MkdirAll(path, 0755)
+	}
+	return nil
+}
+
+// writeFile writes data to path in a single transaction.
+func writeFile(fs utilfs.Filesystem, path string, data []byte) (retErr error) {
+	// Create a temporary file in the base directory of `path` with a prefix.
+	tmpFile, err := fs.TempFile(filepath.Dir(path), tmpPrefix)
+	if err != nil {
+		return err
+	}
+
+	tmpPath := tmpFile.Name()
+
+	defer func() {
+		// Close the file.
+		if err := tmpFile.Close(); err != nil {
+			if retErr == nil {
+				retErr = err
+			} else {
+				retErr = fmt.Errorf("failed to close temp file after error %v", retErr)
+			}
+		}
+
+		// Clean up the temp file on error.
+		if retErr != nil && tmpPath != "" {
+			if err := removePath(fs, tmpPath); err != nil {
+				retErr = fmt.Errorf("failed to remove the temporary file after error %v", retErr)
```

Comment:
Should we include the error from `removePath` in this message as well?

Reply:
Done.